File: convert_between_1D_and_2D.cpp
#include <iostream>

using namespace std;

// Khởi tạo mảng-2-chiều ngẫu nhiên
int** Init2D(int rows, int cols)
{
    // Cấp phát bộ nhớ cho các hàng
    int** a = new int*[rows];

    // Duyệt mảng a theo hàng
    for (int r = 0; r < rows; r++)
    {
        // Ứng với mỗi hàng, cấp phát bộ nhớ cho các cột
        a[r] = new int[cols];
        
        for (int c = 0; c < cols; c++)
        {
            a[r][c] = rand() % 10;
        }
    }
    
    return a;
}

// Chuyển đổi từ mảng-2-chiều thành mảng-1-chiều
int* Convert_2D_To_1D(int** a, int rows, int cols)
{
    int size = rows * cols;

    int* b = new int[size];

    for (int r = 0; r < rows; r++)
    {
        for (int c = 0; c < cols; c++)
        {
            // Công thức tính vị trí của phần tử nào đó trong mảng-1-chiều
            int i = r * cols + c;
            b[i] = a[r][c];
        }
    }

    return b;
}

// Khởi tạo mảng-1-chiều ngẫu nhiên
int* Init1D(int size)
{
    // Cấp phát bộ nhớ cho các phần tử
    int* a = new int[size];

    for (int i = 0; i < size; i++)
    {
        a[i] = rand() % 10;
    }
    
    return a;
}

// Chuyển đổi từ mảng-1-chiều thành mảng-2-chiều
int** Convert_1D_To_2D(int* a, int size, int rows, int cols)
{
    if (rows * cols == size)
    {
        // Cấp phát bộ nhớ cho các hàng của mảng-2-chiều b        
        int** b = new int*[rows];

        for (int r = 0; r < rows; r++)
        {
            // Ứng với mỗi hàng, cấp phát bộ nhớ cho các cột
            b[r] = new int[cols];
        }

        // Duyệt các phần tử của mảng-1-chiều
        for (int i = 0; i < size; i++)
        {
            // Công thức tính chỉ số hàng và chỉ số cột của phần tử nào đó trong mảng-2-chiều
            int r = i / cols;
            int c = i % cols;

            b[r][c] = a[i];
        }

        return b;
    }
    else
    {
        // Trả về con trỏ rỗng khi kích thước của mảng-1-chiều không thể chia thành kích thước hàng và cột của mảng-2-chiều 
        return nullptr;
    }
}

// In mảng-2-chiều
void Show2D(int** a, int rows, int cols, string message)
{
    cout << '\n' << message << ":\n";

    for (int r = 0; r < rows; r++)
    {
        for (int c = 0; c < cols; c++)
        {
            cout << a[r][c] << ' ';
        }

        cout << endl;
    }
}

// In mảng-1-chiều
void Show1D(int* a, int size, string message)
{
    cout << '\n' << message << ":\n";

    for (int i = 0; i < size; i++)
    {
        cout << a[i] << ' ';
    }
}

int main()
{
    // Bàn cờ có kích 8x8
    int Rows = 4;
    int Cols = 6;

    // Khởi tạo mảng-2-chiều
    int** Arr_2D = Init2D(Rows, Cols);

    // Chuyển đổi thành mảng-1-chiều
    int* Arr_1D = Convert_2D_To_1D(Arr_2D, Rows, Cols);

    Show2D(Arr_2D, Rows, Cols, "Mảng-2-chiều");
    Show1D(Arr_1D, Rows * Cols, "Chuyển đổi thành mảng-1-chiều");

    // Khởi tạo mảng-1-chiều
    int Size = 21;
    int* C_1D = Init1D(Size);

    // Chuyển đổi thành mảng-2-chiều có kích thước 3x7
    Rows = 3;
    Cols = 7;
    int** C_2D = Convert_1D_To_2D(C_1D, Size, Rows, Cols);

    
    Show1D(C_1D, Size, "Mảng-1-chiều");
    Show2D(C_2D, Rows, Cols, "Chuyển đổi thành mảng-2-chiều");

    return 0;
}

File: convert_between_1D_and_2D.py
import random as rd

# Khởi tạo mảng-2-chiều ngẫu nhiên
def init_2d(rows, cols):
    L2 = [[rd.randint(0, 9) for c in range(cols)] for r in range(rows)]
    return L2


# Chuyển đổi từ mảng-2-chiều thành mảng-1-chiều
def convert_2d_to_1d(L2):
    rows = len(L2)
    cols = len(L2[0])
    
    # Khởi tạo số lượng phần tử cho mảng-1-chiều
    L1 = [0 for i in range(rows * cols)]   

    for r in range(rows):
        for c in range(cols):
            # Công thức tính vị trí của phần tử nào đó trong mảng-1-chiều
            i = r * cols + c;
            L1[i] = L2[r][c]
    
    return L1


# Khởi tạo mảng-1-chiều ngẫu nhiên
def init_1d(size):
    L = [rd.randint(0, 9) for i in range(size)]
    return L


# Chuyển đổi từ mảng-1-chiều thành mảng-2-chiều
def convert_1d_to_2d(L1, rows, cols):
    size = len(L1)
        
    if rows * cols == size:
        # Khởi tạo mảng-2-chiều
        L2 = [[0 for c in range(cols)] for r in range(rows)]

        # Duyệt các phần tử của mảng-1-chiều
        for i in range(size):
            # ông thức tính chỉ số hàng và chỉ số cột của phần tử nào đó trong mảng-2-chiều
            r = i // cols
            c = i % cols
            L2[r][c] = L1[i]
        
        return L2
    else:
        # Trả về mảng rỗng khi kích thước của mảng-1-chiều không thể chia thành kích thước hàng và cột của mảng-2-chiều
        return []


# In mảng-2-chiều
def show_2d(L, message):
    print(f'{message}:')
    for r in range(len(L)):
        print(*L[r], sep=' ')


# In mảng-1-chiều
def show_1d(L, message):
    print(f'{message}:')
    print(*L, sep=' ')


if __name__ == '__main__':
    rows = 4
    cols = 6
    
    # Khởi tạo mảng-2-chiều
    arr_2D = init_2d(rows, cols)

    # Chuyển đổi thành mảng-1-chiều
    arr_1D = convert_2d_to_1d(arr_2D)

    show_2d(arr_2D, 'Mảng-2-chiều')
    show_1d(arr_1D, 'Chuyển đổi thành mảng-1-chiều')

    # Khởi tạo mảng-1-chiều
    size = 21
    c_1D = init_1d(size)

    # Chuyển đổi thành mảng-2-chiều có kích thước 3x7
    rows = 3
    cols = 7
    
    c_2D = convert_1d_to_2d(c_1D, rows, cols)

    show_1d(c_1D, 'Mảng-1-chiều')
    show_2d(c_2D, 'Chuyển đổi thành mảng-2-chiều')